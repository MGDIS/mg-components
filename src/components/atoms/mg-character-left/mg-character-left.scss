mg-character-left {
  display: block;
}

p {
  margin: 0;
}
