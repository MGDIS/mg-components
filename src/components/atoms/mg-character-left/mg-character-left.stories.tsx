import { h } from "@stencil/core";

export default {
  component: 'mg-character-left',
  title: 'Atoms/mg-character-left',
  parameters: {
    docs: {
      source: {
        code: `<mg-character-left
  identifier="identifier"
  characters=""
  maxlength="400"
  template="{counter} characters left."
></mg-character-left>`,
      },
    },
  },
};

const Template = args => <mg-character-left {...args}></mg-character-left>;

export const MgCharacterLeft = Template.bind({});
MgCharacterLeft.args = {
  identifier: 'identifier',
  characters: '',
  maxlength: 400,
  template: '{counter} characters left.',
};
