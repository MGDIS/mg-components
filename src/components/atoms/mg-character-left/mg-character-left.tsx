import { Component, Host, h, Prop, Watch } from '@stencil/core';
import locale from '../../../locales';

@Component({
  tag: 'mg-character-left',
  styleUrl: 'mg-character-left.scss',
  scoped: true,
})
export class MgCharacterLeft {

  /**
   * Internal
   */
  private mustacheCounter = '{counter}';

  /**
   * Sets an `id` attribute.
   * Needed by the input for accessibility `arai-decribedby`.
   */
  @Prop() identifier: string;

   /**
   * Template to display remaining characters.
   * Must have {counter} inside
   */
  @Prop() template: string = locale.nbCharLeft;
  @Watch('template')
  validateTemplate(newValue: string) {
    if (typeof newValue !== 'string' || newValue === '' || newValue.indexOf(this.mustacheCounter) === -1) {
      throw new Error(`<mg-character-left> prop "template" must contain "${this.mustacheCounter}".`);
    };
  }

  /**
   * Sets the characters to count
   */
  @Prop() characters: string = '';

  /**
   * Add maximum length
   */
  @Prop() maxlength!: number;
  @Watch('maxlength')
  validateMaxlength(newValue: number) {
    if (typeof newValue !== 'number') {
      throw new Error('<mg-character-left> prop "maxlength" is required.')
    };
  }

  /**
   * Calculate number character left
   * @returns {string}
   */
  private getMessage = () => {
    return this.template.replace(this.mustacheCounter, `<strong>${this.maxlength - this.characters.length}</strong>`)
  }

  /**
   * Check if props are well configured on init
   */
  componentWillLoad() {
    this.validateTemplate(this.template);
    this.validateMaxlength(this.maxlength);
  }

  /**
   * Render
   */
  render() {
    return (
      <Host>
        <p id={this.identifier} innerHTML={this.getMessage()}></p>
      </Host>
    );
  }

}
