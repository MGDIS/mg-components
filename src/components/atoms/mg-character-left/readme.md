## Design

### Message order

When messages are displayed under the field the order is as follows:

1. number of characters remaining
2. input help
3. error

### Indication of the number of characters remaining

When the focus is on the input field, a message is displayed underneath. The message is: "`<nbCar>` characters available", `<nbCar>` being a counter that updates dynamically according to the input.
When the focus is no longer on the input field, the message disappears.

Translated with www.DeepL.com/Translator (free version)

![](./mg-character-left/doc/img/nombre-caracteres-restants.png)

<!-- Auto Generated Below -->


## Properties

| Property                 | Attribute    | Description                                                                      | Type     | Default               |
| ------------------------ | ------------ | -------------------------------------------------------------------------------- | -------- | --------------------- |
| `characters`             | `characters` | Sets the characters to count                                                     | `string` | `''`                  |
| `identifier`             | `identifier` | Sets an `id` attribute. Needed by the input for accessibility `aria-decribedby`. | `string` | `undefined`           |
| `maxlength` _(required)_ | `maxlength`  | Add maximum length                                                               | `number` | `undefined`           |
| `template`               | `template`   | Template to display remaining characters. Must have {counter} inside             | `string` | `messages.nbCharLeft` |


## Dependencies

### Used by

 - [mg-input-checkbox](../../molecules/inputs/mg-input-checkbox)
 - [mg-input-date](../../molecules/inputs/mg-input-date)
 - [mg-input-numeric](../../molecules/inputs/mg-input-numeric)
 - [mg-input-password](../../molecules/inputs/mg-input-password)
 - [mg-input-radio](../../molecules/inputs/mg-input-radio)
 - [mg-input-select](../../molecules/inputs/mg-input-select)
 - [mg-input-text](../../molecules/inputs/mg-input-text)
 - [mg-input-textarea](../../molecules/inputs/mg-input-textarea)
 - [mg-input-toggle](../../molecules/inputs/mg-input-toggle)

### Graph
```mermaid
graph TD;
  mg-input-checkbox --> mg-character-left
  mg-input-date --> mg-character-left
  mg-input-numeric --> mg-character-left
  mg-input-password --> mg-character-left
  mg-input-radio --> mg-character-left
  mg-input-select --> mg-character-left
  mg-input-text --> mg-character-left
  mg-input-textarea --> mg-character-left
  mg-input-toggle --> mg-character-left
  style mg-character-left fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
