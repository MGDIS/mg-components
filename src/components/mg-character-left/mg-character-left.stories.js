export default {
  component: 'mg-character-left',
  title: 'Atoms/mg-character-left',
};

const Template = args => <mg-character-left {...args}>{args.slot}</mg-character-left>;

export const MgCharacterLeft = Template.bind({});
MgCharacterLeft.args = {
  reference: 'test-label',
  characters: '',
  maxlength: 400,
  template: '{counter} characters left.',
};
