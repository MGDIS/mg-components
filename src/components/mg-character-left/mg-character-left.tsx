import { Component, Host, h, Prop, Watch } from '@stencil/core';

@Component({
  tag: 'mg-character-left',
  styleUrl: 'mg-character-left.scss',
  scoped: true,
})
export class MgCharacterLeft {

  /**
   * Internal
   */
  private mustacheCounter = '{counter}';

  /**
   * Sets an `id` element.
   * Needed by the input for accessibility `arai-decribedby`.
   */
  @Prop() reference: string;

   /**
   * Template to display remaining characters.
   * Must have {counter} inside
   */
  @Prop() template: string = '{counter} caract√®res disponibles.';
  @Watch('template')
  validateTemplate(newValue: string) {
    if (typeof newValue !== 'string' || newValue === '' || newValue.indexOf(this.mustacheCounter) === -1) {
      throw new Error(`<mg-character-left> prop "template" must contain "${this.mustacheCounter}"`);
    };
  }

  /**
   * Sets the characters to count
   */
  @Prop() characters: string = '';

  /**
   * Add maximum length
   */
  @Prop() maxlength!: number;
  @Watch('maxlength')
  validateMaxlength(newValue: number) {
    if (typeof newValue !== 'number') {
      throw new Error('<mg-character-left> prop "maxlength" is required')
    };
  }

  /**
   * Calculate number character left
   * @returns {string}
   */
  private getMessage = () => {
    return this.template.replace(this.mustacheCounter, `<strong>${this.maxlength - this.characters.length}</strong>`)
  }

  /**
   * Check if props are well configured on init
   */
  componentWillLoad() {
    this.validateTemplate(this.template);
    this.validateMaxlength(this.maxlength);
  }

  /**
   * Render
   */
  render() {
    return (
      <Host>
        <p id={this.reference} innerHTML={this.getMessage()}></p>
      </Host>
    );
  }

}
