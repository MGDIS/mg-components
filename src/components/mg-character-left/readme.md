# mg-character-left



<!-- Auto Generated Below -->


## Properties

| Property                 | Attribute    | Description                                                                    | Type     | Default                               |
| ------------------------ | ------------ | ------------------------------------------------------------------------------ | -------- | ------------------------------------- |
| `characters`             | `characters` | Sets the characters to count                                                   | `string` | `''`                                  |
| `maxlength` _(required)_ | `maxlength`  | Add maximum length                                                             | `number` | `undefined`                           |
| `reference`              | `reference`  | Sets an `id` element. Needed by the input for accessibility `arai-decribedby`. | `string` | `undefined`                           |
| `template`               | `template`   | Template to display remaining characters. Must have {counter} inside           | `string` | `'{counter} caractÃ¨res disponibles.'` |


## Dependencies

### Used by

 - [mg-input-text](../mg-input-text)

### Graph
```mermaid
graph TD;
  mg-input-text --> mg-character-left
  style mg-character-left fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
