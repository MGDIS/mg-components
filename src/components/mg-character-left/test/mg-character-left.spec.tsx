import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgCharacterLeft } from '../mg-character-left';

const getPage = (args) => newSpecPage({
  components: [MgCharacterLeft],
  template: () => (<mg-character-left {...args}></mg-character-left>),
});

describe.each(['blu', 'blublu', 'blu blu blu blu blu'])('mg-character-left for %s', (characters) => {
  test.each([123, 234, 200])('renders', async (maxlength) => {
    const expected = maxlength - characters.length;
    const page = await getPage({characters, maxlength});
    expect(page.root).toEqualHtml(`
      <mg-character-left>
      <p>Reste ${expected} caractères à saisir</p>
      </mg-character-left>
    `);
  });
});
