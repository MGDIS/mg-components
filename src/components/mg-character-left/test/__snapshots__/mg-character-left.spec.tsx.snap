// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-character-left Should render nb char left for blu blu blu blu blu renders 1`] = `
<mg-character-left>
  <p>
    <strong>
      104
    </strong>
    caractères disponibles.
  </p>
</mg-character-left>
`;

exports[`mg-character-left Should render nb char left for blu blu blu blu blu renders 2`] = `
<mg-character-left>
  <p>
    <strong>
      215
    </strong>
    caractères disponibles.
  </p>
</mg-character-left>
`;

exports[`mg-character-left Should render nb char left for blu blu blu blu blu renders 3`] = `
<mg-character-left>
  <p>
    <strong>
      181
    </strong>
    caractères disponibles.
  </p>
</mg-character-left>
`;

exports[`mg-character-left Should render nb char left for blu renders 1`] = `
<mg-character-left>
  <p>
    <strong>
      120
    </strong>
    caractères disponibles.
  </p>
</mg-character-left>
`;

exports[`mg-character-left Should render nb char left for blu renders 2`] = `
<mg-character-left>
  <p>
    <strong>
      231
    </strong>
    caractères disponibles.
  </p>
</mg-character-left>
`;

exports[`mg-character-left Should render nb char left for blu renders 3`] = `
<mg-character-left>
  <p>
    <strong>
      197
    </strong>
    caractères disponibles.
  </p>
</mg-character-left>
`;

exports[`mg-character-left Should render nb char left for blublu renders 1`] = `
<mg-character-left>
  <p>
    <strong>
      117
    </strong>
    caractères disponibles.
  </p>
</mg-character-left>
`;

exports[`mg-character-left Should render nb char left for blublu renders 2`] = `
<mg-character-left>
  <p>
    <strong>
      228
    </strong>
    caractères disponibles.
  </p>
</mg-character-left>
`;

exports[`mg-character-left Should render nb char left for blublu renders 3`] = `
<mg-character-left>
  <p>
    <strong>
      194
    </strong>
    caractères disponibles.
  </p>
</mg-character-left>
`;
